x-pack/plugins/maps/common/migrations/references.js

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+// Can not use public Layer classes to extract references since this logic must run in both client and server.
+
+import _ from 'lodash';
+
+export function extractReferences({ attributes, references = [] }) {
+  if (!attributes.layerListJSON) {
+    return { attributes, references };
+  }
+
+  const extractedReferences = [];
+
+  const layerList = JSON.parse(attributes.layerListJSON);
+  layerList.forEach((layer, layerIndex) => {
+
+    // Extract index-pattern references from source descriptor
+    if (_.has(layer, 'sourceDescriptor.indexPatternId')) {
```

Comment:
iso relying on convention, I'd do an explicit type-check here on the sources that actually have an `indexPatternId` property in their descriptor. We know exactly when this happens (`ES_GEO_GRID`, `ES_SEARCH`). The type-name will tie this more explicitly to the code in `public/shared/sources` as well.